atomic_reactor/plugins/pre_yum_proxy.py

```diff
@@ -0,0 +1,162 @@
+            # SCRATCH_FROM are exluded from parent_images .
+            self.all_yum_repo_files[parent_image_id] = self._get_yum_repos(parent_image_id)
+
+        self._generate_new_yum_repo_files()
+        self._update_dockerfile()
+
+    def _get_yum_repos(self, image_id):
+        # TODO: This needs a retry
+        dest_path = os.path.join(self.workdir, 'original', image_id)
+        try:
+            os.makedirs(dest_path)
+        except OSError:
+            # Ignore if dirs have already been created. Python 2 doesn't suport exists_ok param
+            pass
+        container_id = self.tasker.copy_file(
+            image_id,
+            src_path='/etc/yum.repos.d',
+            dest_path=dest_path,
+            create_kwargs={'user': 'root'},
+        )
+        self._container_ids.append(container_id)
+        return self._parse_yum_repos(dest_path)
+
+    def _parse_yum_repos(self, dest_path):
+        dest_path = os.path.join(dest_path, 'yum.repos.d')
+        yum_repo_files = {}
+        for repo_file_path in os.listdir(dest_path):
+            repo_file_content = ConfigParser()
+            repo_file_content.read(os.path.join(dest_path, repo_file_path))
+            yum_repo_files[repo_file_path] = repo_file_content
+        return yum_repo_files
+
+    def _generate_new_yum_repo_files(self):
+        for parent_image_id, yum_repo_files in self.all_yum_repo_files.items():
+            for yum_repo_file_name, yum_repo_file_content in yum_repo_files.items():
+                rewrite_file = False
+                for yum_repo_name in yum_repo_file_content.sections():
+                    baseurl = yum_repo_file_content.get(yum_repo_name, 'baseurl')
+                    if not baseurl:
+                        continue
+                    parts = urlsplit(baseurl)
+                    if parts.netloc in self.yum_proxies:
+                        # TODO: Do we also need to replace URL schema?
+                        proxied_baseurl = urlunsplit(
+                            (parts[0], self.yum_proxies[parts.netloc]) + parts[2:])
+                        yum_repo_file_content.set(yum_repo_name, 'baseurl', proxied_baseurl)
+                        rewrite_file = True
+
+                if rewrite_file:
+                    dest_path = os.path.join(self.workdir, 'updated', parent_image_id,
+                                             yum_repo_file_name)
+                    self._write_yum_repo_file(dest_path, yum_repo_file_content)
+                    self.updated_yum_repo_files.setdefault(parent_image_id, [])
+                    self.updated_yum_repo_files[parent_image_id].append(yum_repo_file_name)
+
+    def _write_yum_repo_file(self, dest_path, content):
+        try:
+            os.makedirs(os.path.dirname(dest_path))
+        except OSError:
+            # Ignore if dirs have already been created. Python 2 doesn't suport exists_ok param
+            pass
+        with open(dest_path, 'w') as f:
```

Comment:
Do we need to take special action to preserve the file mode of the original file?